src/main/java/de/numcodex/feasibility_gui_backend/query/broker/direct/DirectBrokerClientCql.java

```
@@ -0,0 +1,186 @@
+    private final IGenericClient fhirClient;
+
+    /**
+     * Creates a new {@link DirectBrokerClientCql} instance that uses the given web client to
+     * communicate with a CQL capable FHIR server instance.
+     *
+     * @param fhirContext A FHIR context.
+     * @param fhirClient A FHIR client, configured with the correct url
+     */
+    public DirectBrokerClientCql(FhirContext fhirContext, IGenericClient fhirClient) {
+        this.fhirContext = Objects.requireNonNull(fhirContext);
+        this.fhirClient = Objects.requireNonNull(fhirClient);
+        listeners = new ArrayList<>();
+        brokerQueries = new HashMap<>();
+        brokerToBackendQueryIdMapping = new HashMap<>();
+    }
+
+    @Override
+    public void publishQuery(String brokerQueryId) throws QueryNotFoundException, IOException {
+        var query = findQuery(brokerQueryId);
+        var queryContent = Optional.ofNullable(query.getQueryDefinition(CQL))
+            .orElseThrow(() -> new IllegalStateException("Query with ID "
+                + query.getQueryId()
+                + " does not contain a query definition for the mandatory type: "
+                + CQL
+            ));
+        try {
+            updateQueryStatus(query, QueryStatus.EXECUTING);
+            var measureUri = createMeasureAndLibrary(queryContent);
+            var report = evaluateMeasure(measureUri);
+            var resultCount = report.getGroupFirstRep().getPopulationFirstRep().getCount();
+            query.registerSiteResults(SITE_1_ID, obfuscateResultCount ? obfuscate(resultCount) : resultCount);
+            updateQueryStatus(query, COMPLETED);
+        } catch (Exception e) {
+            updateQueryStatus(query, FAILED);
+            throw new IOException("An error occurred while publishing the query with ID: " + query.getQueryId(), e);
+        }
+    }
+
+    @Override
+    public String getSiteName(String siteId) {
+        return siteId.equals(SITE_1_ID) ? SITE_1_NAME : "";
```

Shouldn't `getSiteName` throw an `SiteNotFoundException` instead of returning an empty string?

---

That implementation can be moved into `DirectBrokerClient`. Also reversing equals saves us from an NPE here.